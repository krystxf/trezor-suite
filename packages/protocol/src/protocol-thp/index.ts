export * from './decode';
export * from './encode';
export * from './messages';
export { getCpaceHostKeys, getShareSecret, handleHandshakeInitResponse } from './crypto/pairing';
export { ThpProtocolState } from './ThpProtocolState';
export { getCurve25519KeyPair } from './crypto/curve25519';

export type { ThpMessageType, ThpDeviceProperties } from './messages';
export const name = 'thp';
