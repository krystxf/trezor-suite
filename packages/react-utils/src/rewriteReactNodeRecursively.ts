import { ReactNode, isValidElement, cloneElement, Key } from 'react';

type RewriteReactNodeRecursivelyCallback = (stringSubNode: string) => string;

/**
 * Recursively crawls ReactNode and applies callback on all string subnodes
 * @param node ReactNode to be recursively crawled & modified
 * @param callback function to be applied on all string subnodes
 * @param reactKey optional React key applied on the root node, if it is a ReactElement. Used internally.
 * @returns modified ReactNode
 */
export const rewriteReactNodeRecursively = (
    node: ReactNode,
    callback: RewriteReactNodeRecursivelyCallback,
    reactKey?: Key,
): ReactNode => {
    if (typeof node === 'string') return callback(node);
    if (typeof node === 'number') return callback(node.toString());

    if (Array.isArray(node)) {
        return node.map((child, index) => {
            const newReactKey: Key | undefined = isValidElement(child)
                ? child.key ?? index
                : undefined;

            return rewriteReactNodeRecursively(child, callback, newReactKey);
        });
    }

    if (isValidElement(node)) {
        const rewrittenChildren = rewriteReactNodeRecursively(node.props.children, callback);

        return cloneElement(node, { ...node.props, key: reactKey, children: rewrittenChildren });
    }

    return node;
};
