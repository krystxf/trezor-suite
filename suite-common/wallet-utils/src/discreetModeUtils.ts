export const DISCREET_PLACEHOLDER = '####';

const numericalSubstringRegex = /[\d\-\.,]*\d+[\.\d]*/g;

/**
 * Search the input for a numerical substring and replace it with DISCREET_PLACEHOLDER.
 * @param value whole string value, usually number with symbol
 * @returns redacted string
 */
export const redactNumericalSubstring = (value: string): string =>
    value.replace(numericalSubstringRegex, DISCREET_PLACEHOLDER);
