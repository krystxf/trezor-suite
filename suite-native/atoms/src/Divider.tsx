import { G } from '@mobily/ts-belt';

import {
    NativeStyleObject,
    mergeNativeStyleObjects,
    prepareNativeStyle,
    useNativeStyles,
} from '@trezor/styles';

import { Box, BoxProps } from './Box';

type DividerProps = Omit<BoxProps, 'style'> & {
    style?: NativeStyleObject;
};

const dividerStyle = prepareNativeStyle(utils => ({
    borderBottomWidth: utils.borders.widths.small,
    borderBottomColor: utils.colors.borderElevation1,
    flex: 1,
}));

export const Divider = ({ style, ...props }: DividerProps) => {
    const { applyStyle } = useNativeStyles();

    const defaultStyle = applyStyle(dividerStyle);

    const mergedStyle = G.isNotNullable(style)
        ? mergeNativeStyleObjects([defaultStyle, style])
        : defaultStyle;

    return <Box style={mergedStyle} {...props} />;
};
