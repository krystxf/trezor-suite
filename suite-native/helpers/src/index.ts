export * from './hooks/useCopyToClipboard';
export * from './splitAddressToChunks';
export * from './constants';
