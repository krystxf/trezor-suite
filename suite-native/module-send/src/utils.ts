import { SendOutputFieldName } from './sendOutputsFormSchema';

export const sendAmountTransformer = (value: string) =>
    value
        .replace(/[^0-9\.]/g, '') // remove all non-numeric characters
        .replace(/^\./g, '') // remove '.' symbol if it is not preceded by number
        .replace(/(?<=\..*)\./g, '') // keep only first appearance of the '.' symbol
        .replace(/(?<=^0+)0/g, ''); // remove all leading zeros except the first one

export const getOutputFieldName = <TField extends SendOutputFieldName>(
    index: number,
    field: TField,
): `outputs.${number}.${TField}` => `outputs.${index}.${field}`;
