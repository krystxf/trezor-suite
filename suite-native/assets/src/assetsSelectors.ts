import { memoize } from 'proxy-memoize';
import { A, F, pipe } from '@mobily/ts-belt';

import { BigNumber } from '@trezor/utils/src/bigNumber';
import { networks, NetworkSymbol } from '@suite-common/wallet-config';
import {
    AccountsRootState,
    DeviceRootState,
    FiatRatesRootState,
    selectVisibleDeviceAccounts,
    selectFiatRatesByFiatRateKey,
} from '@suite-common/wallet-core';
import { getFiatRateKey, toFiatCurrency } from '@suite-common/wallet-utils';
import { selectFiatCurrencyCode, SettingsSliceRootState } from '@suite-native/settings';
import { discoverySupportedNetworks } from '@suite-native/config';

type Assets = Partial<Record<NetworkSymbol, string[]>>;
type FormattedAssets = Partial<Record<NetworkSymbol, BigNumber>>;

export interface AssetType {
    symbol: NetworkSymbol;
    network: (typeof networks)[NetworkSymbol];
    assetBalance: BigNumber;
    fiatBalance: string;
}

type AssetsRootState = AccountsRootState & FiatRatesRootState & SettingsSliceRootState;

const sumBalance = (balances: string[]): BigNumber =>
    balances.reduce((prev, balance) => prev.plus(balance), new BigNumber(0));

export const selectDeviceBalancesPerNetwork = memoize(
    (state: AssetsRootState & DeviceRootState): FormattedAssets => {
        const accounts = selectVisibleDeviceAccounts(state);

        const assets: Assets = {};
        accounts.forEach(account => {
            if (!assets[account.symbol]) {
                assets[account.symbol] = [];
            }
            assets[account.symbol]?.push(account.formattedBalance);
        });

        const formattedNetworkAssets: FormattedAssets = {};
        const assetKeys = Object.keys(assets) as NetworkSymbol[];
        assetKeys.forEach((asset: NetworkSymbol) => {
            const balances = assets[asset] ?? [];
            formattedNetworkAssets[asset] = sumBalance(balances);
        });

        return formattedNetworkAssets;
    },
);

export const selectDeviceAssetsWithBalances = memoize(
    (state: AssetsRootState & DeviceRootState) => {
        const deviceBalancesPerNetwork = selectDeviceBalancesPerNetwork(state);
        const deviceNetworksWithAssets = Object.keys(deviceBalancesPerNetwork) as NetworkSymbol[];

        const fiatCurrencyCode = selectFiatCurrencyCode(state);

        return pipe(
            deviceNetworksWithAssets,
            A.sort((a, b) => {
                const aOrder = discoverySupportedNetworks.indexOf(a) ?? Number.MAX_SAFE_INTEGER;
                const bOrder = discoverySupportedNetworks.indexOf(b) ?? Number.MAX_SAFE_INTEGER;

                return aOrder - bOrder;
            }),
            A.map((networkSymbol: NetworkSymbol) => {
                const fiatRateKey = getFiatRateKey(networkSymbol, fiatCurrencyCode);
                const fiatRate = selectFiatRatesByFiatRateKey(state, fiatRateKey);

                // Note: This shouldn't be happening in a selector but rather in component itself.
                // In future, we will probably have something like `CryptoAmountToFiatFormatter` in component just using value sent from this selector.
                const fiatBalance =
                    toFiatCurrency(
                        deviceBalancesPerNetwork[networkSymbol]?.toString() ?? '0',
                        fiatRate?.rate,
                    ) ?? '0';

                const network = networks[networkSymbol];

                const asset: AssetType = {
                    symbol: networkSymbol,
                    network,
                    assetBalance: deviceBalancesPerNetwork[networkSymbol] ?? new BigNumber(0),
                    fiatBalance,
                };

                return asset;
            }),
            A.filter(data => data !== undefined),
            F.toMutable,
        );
    },
);
