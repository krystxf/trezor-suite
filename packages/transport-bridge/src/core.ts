import { WebUSB } from 'usb';

import {
    v1 as protocolV1,
    v2 as protocolV2,
    thp as protocolThp,
    bridge as protocolBridge,
    TransportProtocol,
} from '@trezor/protocol';
import { receive as receiveUtil } from '@trezor/transport/src/utils/receive';
import { sendThpMessage, receiveThpMessage } from '@trezor/transport/src/thp';
import { createChunks, sendChunks } from '@trezor/transport/src/utils/send';
import { SessionsBackground } from '@trezor/transport/src/sessions/background';
import { SessionsClient } from '@trezor/transport/src/sessions/client';
import { UsbApi } from '@trezor/transport/src/api/usb';
import { UdpApi } from '@trezor/transport/src/api/udp';
import { AcquireInput, ReleaseInput } from '@trezor/transport/src/transports/abstract';
import { Session, BridgeProtocolMessage } from '@trezor/transport/src/types';
import { createProtocolMessage } from '@trezor/transport/src/utils/bridgeProtocolMessage';
import { Log } from '@trezor/utils';
import { AbstractApi } from '@trezor/transport/src/api/abstract';

export const createCore = (apiArg: 'usb' | 'udp' | AbstractApi, logger?: Log) => {
    let api: AbstractApi;

    const abortController = new AbortController();
    const sessionsBackground = new SessionsBackground({ signal: abortController.signal });

    const sessionsClient = new SessionsClient({
        requestFn: args => sessionsBackground.handleMessage(args),
        registerBackgroundCallbacks: () => {},
    });

    sessionsBackground.on('descriptors', descriptors => {
        sessionsClient.emit('descriptors', descriptors);
    });

    if (typeof apiArg === 'string') {
        api =
            apiArg === 'udp'
                ? new UdpApi({ logger })
                : new UsbApi({
                      logger,
                      usbInterface: new WebUSB({
                          allowAllDevices: true, // return all devices, not only authorized
                      }),

                      // todo: possibly only for windows
                      forceReadSerialOnConnect: true,
                  });
    } else {
        api = apiArg;
    }

    api.listen();

    // whenever low-level api reports changes to descriptors, report them to sessions module
    api.on('transport-interface-change', descriptors => {
        logger?.debug(`core: transport-interface-change ${JSON.stringify(descriptors)}`);
        sessionsClient.enumerateDone({ descriptors });
    });

    const writeUtil = async ({
        path,
        data,
        signal,
        protocol,
    }: {
        path: string;
        data: string;
        signal: AbortSignal;
        protocol: TransportProtocol;
    }) => {
        logger?.debug(`core: writeUtil protocol ${protocol.name}`);
        const buffer = Buffer.from(data, 'hex');
        let encodedMessage;
        let chunkHeader;
        if (protocol.name === 'bridge') {
            const { messageType, payload } = protocolBridge.decode(buffer);
            encodedMessage = protocolV1.encode(payload, { messageType });
            chunkHeader = protocolV1.getChunkHeader(encodedMessage);
        } else {
            encodedMessage = buffer;
            chunkHeader = protocol.getChunkHeader(encodedMessage);
        }

        const chunks = createChunks(encodedMessage, chunkHeader, api.chunkSize);
        const apiWrite = (chunk: Buffer) => api.write(path, chunk, signal);
        const sendResult = await sendChunks(chunks, apiWrite);

        return sendResult;
    };

    const readUtil = async ({
        path,
        signal,
        protocol,
    }: {
        path: string;
        signal: AbortSignal;
        protocol: TransportProtocol;
    }) => {
        logger?.debug(`core: readUtil protocol ${protocol.name}`);
        try {
            const receiveProtocol = protocol.name === 'bridge' ? protocolV1 : protocol;
            const { messageType, payload } = await receiveUtil(() => {
                logger?.debug(`core: readUtil: api.read: reading next chunk`);

                return api.read(path, signal).then(result => {
                    if (result.success) {
                        logger?.debug(
                            `core: readUtil partial result: byteLength: ${result.payload.byteLength}`,
                        );

                        return result.payload;
                    }
                    logger?.debug(`core: readUtil partial result: error: ${result.error}`);
                    throw new Error(result.error);
                });
            }, receiveProtocol);

            logger?.debug(
                `core: readUtil result: messageType: ${messageType} byteLength: ${payload?.byteLength}`,
            );

            return {
                success: true as const,
                payload: protocol.encode(payload, { messageType }).toString('hex'),
            };
        } catch (err) {
            logger?.debug(`core: readUtil catch: ${err.message}`);

            return { success: false as const, error: err.message as string };
        }
    };

    const enumerate = async ({ signal }: { signal: AbortSignal }) => {
        await sessionsClient.enumerateIntent();
        const enumerateResult = await api.enumerate(signal);

        if (!enumerateResult.success) {
            return enumerateResult;
        }

        const enumerateDoneResponse = await sessionsClient.enumerateDone({
            descriptors: enumerateResult.payload,
        });

        return enumerateDoneResponse;
    };

    const acquire = async (
        acquireInput: Omit<AcquireInput, 'previous'> & {
            previous: Session | 'null';
            signal: AbortSignal;
        },
    ) => {
        const acquireIntentResult = await sessionsClient.acquireIntent({
            path: acquireInput.path,
            previous: acquireInput.previous === 'null' ? null : acquireInput.previous,
        });
        if (!acquireIntentResult.success) {
            return acquireIntentResult;
        }

        const openDeviceResult = await api.openDevice(acquireInput.path, true, acquireInput.signal);
        logger?.debug(`core: openDevice: result: ${JSON.stringify(openDeviceResult)}`);

        if (!openDeviceResult.success) {
            return openDeviceResult;
        }
        await sessionsClient.acquireDone({ path: acquireInput.path });

        return acquireIntentResult;
    };

    const release = async ({ session }: Omit<ReleaseInput, 'path'>) => {
        await sessionsClient.releaseIntent({ session });

        const sessionsResult = await sessionsClient.getPathBySession({
            session,
        });

        if (!sessionsResult.success) {
            return sessionsResult;
        }

        const closeRes = await api.closeDevice(sessionsResult.payload.path);

        if (!closeRes.success) {
            logger?.error(`core: release: api.closeDevice error: ${closeRes.error}`);
        }

        return sessionsClient.releaseDone({ path: sessionsResult.payload.path });
    };

    const getProtocol = (name: BridgeProtocolMessage['protocol']) => {
        if (name === 'v1') {
            return protocolV1;
        }

        if (name === 'v2') {
            return protocolV2;
        }

        return protocolBridge;
    };

    const createProtocolMessageResponse = (
        protocol: BridgeProtocolMessage['protocol'],
        response: Awaited<ReturnType<typeof readUtil>> | Awaited<ReturnType<typeof writeUtil>>,
    ) => {
        if (response.success) {
            const body = 'payload' in response ? response.payload : '';

            return {
                ...response,
                payload: createProtocolMessage(body, protocol),
            };
        }

        return response;
    };
    // const getRequestBody = (data: string, protocol?: TransportProtocol) => {
    //     if (protocol?.name === 'v2') {
    //         try {
    //             const json = JSON.parse(data);
    //             const protocolState = new protocolThp.ThpProtocolState();
    //             protocolState.deserialize(json.state);

    //             return {
    //                 protocolState,
    //                 body: json.body,
    //             };
    //         } catch (e) {
    //             logger?.debug(`THP getRequestBody error: ${e} ${data}`);
    //             // TODO: break communication and return success: false
    //         }
    //     }

    //     return {
    //         protocolState: undefined,
    //         body: data,
    //     };
    // };

    const call = async ({
        session,
        data,
        signal,
        protocol: protocolName,
        state,
    }: BridgeProtocolMessage & {
        session: Session;
        signal: AbortSignal;
    }) => {
        logger?.debug(`core: call: session: ${session} ${protocolName}`);
        const sessionsResult = await sessionsClient.getPathBySession({
            session,
        });
        if (!sessionsResult.success) {
            logger?.error(`core: call: retrieving path error: ${sessionsResult.error}`);

            return sessionsResult;
        }
        const protocol = getProtocol(protocolName);
        let protocolState;
        if (state) {
            protocolState = new protocolThp.ThpProtocolState();
            protocolState.deserialize(state);
        }

        const { path } = sessionsResult.payload;
        logger?.debug(`core: call: retrieved path ${path} for session ${session}`);

        const openResult = await api.openDevice(path, false, signal);

        if (!openResult.success) {
            logger?.error(`core: call: api.openDevice error: ${openResult.error}`);

            return openResult;
        }
        logger?.debug(`core: call: api.openDevice done`);

        if (protocol.name === 'v2') {
            const b = Buffer.from(data, 'hex');
            const chunks = createChunks(b, protocol.getChunkHeader(b), api.chunkSize);
            protocolState?.setChannel(b.subarray(1, 3));

            const apiWrite = (chunk: Buffer, attemptSignal?: AbortSignal) =>
                api.write(path, chunk, attemptSignal || signal);
            const apiRead = (attemptSignal?: AbortSignal) =>
                api.read(path, attemptSignal || signal);

            await sendThpMessage({
                protocolState,
                chunks,
                apiWrite,
                apiRead,
                signal,
            });

            const message = await receiveThpMessage({
                protocolState,
                apiWrite,
                apiRead,
                signal,
            });

            return createProtocolMessageResponse(protocol.name, {
                success: true,
                payload: protocol.encode(message.payload, message).toString('hex'),
            });
        }

        const writeResult = await writeUtil({ path, data, protocol, signal });
        if (!writeResult.success) {
            logger?.error(`core: call: writeUtil ${writeResult.error}`);

            return writeResult;
        }
        logger?.debug('core: call: readUtil');

        const readResult = await readUtil({ path, signal, protocol });

        return createProtocolMessageResponse(protocolName, readResult);
        // if (!readResult.success) {
        //     return readResult;
        // }

        // if (protocol.name !== 'bridge') {
        //     const buffer = Buffer.from(readResult.payload, 'hex');
        //     const thpMessage = protocol.decode(buffer);
        //     const isThpAck = protocolThp.decodeAck(thpMessage);

        //     if (isThpAck) {
        //         protocolState?.updateSyncBit('send');
        //         // protocolState?.updateNonce('send');

        //         console.warn('ACK in response, reading again');
        //         const realResult = await readUtil({ path, protocol, signal });
        //         if (!realResult.success) {
        //             return realResult;
        //         }
        //         // protocolState?.updateNonce('recv');

        //         console.warn('Send ThpReadAck');
        //         const chunk = protocolThp.encodeAck(Buffer.from(realResult.payload, 'hex'));

        //         const ackResult = await writeUtil({
        //             path,
        //             data: chunk.toString('hex'),
        //             signal,
        //             protocol,
        //         });
        //         if (!ackResult.success) {
        //             return ackResult;
        //         }
        //         protocolState?.updateSyncBit('recv');

        //         return {
        //             ...realResult,
        //             payload: JSON.stringify({
        //                 body: realResult.payload,
        //                 state: protocolState?.serialize(),
        //             }),
        //         };
        //     } else {
        //         return {
        //             ...readResult,
        //             payload: JSON.stringify({
        //                 body: readResult.payload,
        //                 state: protocolState?.serialize(),
        //             }),
        //         };
        //     }
        // }

        // return readResult;
    };

    const send = async ({
        session,
        data,
        signal,
        protocol: protocolName,
        state,
    }: BridgeProtocolMessage & {
        session: Session;
        signal: AbortSignal;
    }) => {
        const sessionsResult = await sessionsClient.getPathBySession({
            session,
        });

        if (!sessionsResult.success) {
            return sessionsResult;
        }
        const { path } = sessionsResult.payload;
        const protocol = getProtocol(protocolName);
        if (state) {
            const protocolState = new protocolThp.ThpProtocolState();
            protocolState.deserialize(state);
        }

        const openResult = await api.openDevice(path, false, signal);
        if (!openResult.success) {
            return openResult;
        }

        if (protocol.name === 'v2') {
            const writeResult = await writeUtil({ path, data, signal, protocol });
            if (!writeResult.success) {
                return writeResult;
            }
            // // Read ACK
            // const readResult = await readUtil({ path, signal, protocol });
            // if (!readResult.success) {
            //     return readResult;
            // }

            return createProtocolMessageResponse(protocolName, writeResult);
        }

        const writeResult = await writeUtil({ path, data, signal, protocol });

        return createProtocolMessageResponse(protocolName, writeResult);
    };

    const receive = async ({
        session,
        signal,
        protocol: protocolName,
        state,
    }: BridgeProtocolMessage & {
        session: Session;
        signal: AbortSignal;
    }) => {
        const sessionsResult = await sessionsClient.getPathBySession({
            session,
        });

        if (!sessionsResult.success) {
            return sessionsResult;
        }
        const { path } = sessionsResult.payload;
        const protocol = getProtocol(protocolName);

        const openResult = await api.openDevice(path, false, signal);
        if (!openResult.success) {
            return openResult;
        }

        if (protocol.name === 'v2') {
            const protocolState = new protocolThp.ThpProtocolState();
            if (state) {
                protocolState.deserialize(state);
                protocolState.setExpectedResponse([4, 128, 1000]);
            }

            const apiWrite = (chunk: Buffer, attemptSignal?: AbortSignal) =>
                api.write(path, chunk, attemptSignal || signal);
            const apiRead = (attemptSignal?: AbortSignal) =>
                api.read(path, attemptSignal || signal);

            try {
                const message = await receiveThpMessage({
                    protocolState,
                    apiWrite,
                    apiRead,
                    signal,
                });

                return createProtocolMessageResponse(protocolName, message);
            } catch (e) {
                return {
                    success: false,
                    error: e.message,
                };
            }
        }

        const readResult = await readUtil({ path, signal, protocol });

        return createProtocolMessageResponse(protocolName, readResult);
    };

    const dispose = () => {
        abortController.abort();
        api.dispose();
        sessionsClient.dispose();
    };

    return {
        enumerate,
        acquire,
        release,
        call,
        send,
        receive,
        dispose,
        sessionsClient,
    };
};
