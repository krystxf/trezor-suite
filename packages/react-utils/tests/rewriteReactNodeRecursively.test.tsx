import { rewriteReactNodeRecursively } from '../src/rewriteReactNodeRecursively';

describe('rewriteReactNodeRecursively', () => {
    const callback = (stringSubNode: string) => stringSubNode.replace(/foo/g, 'bar');

    it('should rewrite a string subnode', () => {
        const inputNode = 'This is foobar.';
        const expectedNode = 'This is barbar.';
        expect(rewriteReactNodeRecursively(inputNode, callback)).toBe(expectedNode);
    });

    it('should cast a number subnode to string and rewrite it', () => {
        const callbackForNumbers = (stringifiedNumber: string) =>
            stringifiedNumber.replace(/34/g, '***');
        const inputNode = 123456;
        const expectedNode = '12***56';
        expect(rewriteReactNodeRecursively(inputNode, callbackForNumbers)).toBe(expectedNode);
    });

    it('should return the node if it is a primitive other than string or number', () => {
        expect(rewriteReactNodeRecursively(null, callback)).toBe(null);
        expect(rewriteReactNodeRecursively(undefined, callback)).toBe(undefined);
        expect(rewriteReactNodeRecursively(true, callback)).toBe(true);
    });

    it('should rewrite all string subnodes in an array', () => {
        const inputNode = ['This is foobar.', 'It has a foo.'];
        const expectedNode = ['This is barbar.', 'It has a bar.'];
        expect(rewriteReactNodeRecursively(inputNode, callback)).toEqual(expectedNode);
    });

    it('should recursively rewrite string subnodes in a ReactElement', () => {
        const inputNode = (
            <span>
                This is foobar, <i>it has a foo</i>.
            </span>
        );
        expect(rewriteReactNodeRecursively(inputNode, callback)).toMatchSnapshot();
    });

    it('should iterate through a collection of ReactNodes, preserving keys or setting them to index if necessary', () => {
        const inputNode = [
            // eslint-disable-next-line react/jsx-key
            <span>Foo does not have key!.</span>,
            <span key="id234">This is foo with key.</span>,
            false,
        ] as const;

        const result = rewriteReactNodeRecursively(inputNode, callback) as typeof inputNode;

        expect(result.length).toBe(3);
        expect(result[0].key).toBe('0');
        expect(result[1].key).toBe('id234');
    });
});
