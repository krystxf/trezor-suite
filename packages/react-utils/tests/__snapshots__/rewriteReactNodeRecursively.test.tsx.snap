// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rewriteReactNodeRecursively should recursively rewrite string subnodes in a ReactElement 1`] = `
<span>
  This is barbar, 
  <i>
    it has a bar
  </i>
  .
</span>
`;
